### PR TITLE
Create users as catalog viewers.

### PR DESCRIPTION
**Pull request by peter-fogg**
@clintonb @bderusha Currently creating a new catalog from the API admin breaks unless all the viewers exist in the catalog service. This fix creates new users if necessary.

### CODE REVIEW

File: course_discovery/apps/api/tests/test_serializers.py [hunk 1]
```diff
@@ -36,6 +37,21 @@ def test_data(self):
...
+            'id': None,
+            'name': 'test new catalog',
+            'query': '*',
+        }
+        self.assertEqual(list(User.objects.filter(username=username)), [])  # pylint: disable=no-member
+        serializer = CatalogSerializer(data=data)
+        self.assertTrue(serializer.is_valid())
+        catalog = serializer.save()
+        self.assertEqual([viewer.username for viewer in catalog.viewers], [username])
+        self.assertEqual(len(User.objects.filter(username=username)), 1)  # pylint: disable=no-member
```
Comment (clintonb) [inline course_discovery/apps/api/tests/test_serializers.py]: `User.objects.filter(username=username).count()`

File: course_discovery/apps/api/tests/test_serializers.py [hunk 2]
```diff
@@ -36,6 +37,21 @@ def test_data(self):
...
 
+    def test_create_new_user(self):
+        username = 'test-user'
+        data = {
+            'viewers': [username],
+            'id': None,
+            'name': 'test new catalog',
+            'query': '*',
+        }
+        self.assertEqual(list(User.objects.filter(username=username)), [])  # pylint: disable=no-member
```
Comment (clintonb) [inline course_discovery/apps/api/tests/test_serializers.py]: `User.objects.filter(username=username).count()`

File: course_discovery/apps/api/serializers.py [hunk 3]
```diff
@@ -93,11 +93,21 @@ class CatalogSerializer(serializers.ModelSerializer):
                                            allow_null=True, allow_empty=True, required=False,
                                            help_text=_('Usernames of users with explicit access to view this catalog'))
 
+    def is_valid(self, **kwargs):
+        # Ensure that the catalog's viewers actually exist in the DB.
+        for username in self.initial_data.get('viewers', ()):  # pylint: disable=no-member
+            User.objects.get_or_create(username=username)
+        return super().is_valid(**kwargs)
```
Comment (clintonb) [inline course_discovery/apps/api/serializers.py]: Call this first. If the request is not valid, we don't want to dirty the database with users that might not get used.

Comment (peter-fogg) [inline course_discovery/apps/api/serializers.py]: The request isn't valid unless the users exist. I suppose we could wrap it in a transaction and roll back if `not super.is_valid()`, but the number of unneeded users is probably going to be low -- if someone attempts to create a catalog and fails, they'll probably just correct the error and try again, so the users will be created eventually anyway.

File: course_discovery/apps/api/tests/test_serializers.py (hunk 2 above)
Comment (peter-fogg) [inline course_discovery/apps/api/tests/test_serializers.py]: Fixed.

File: course_discovery/apps/api/tests/test_serializers.py [hunk 4]
```diff
@@ -36,6 +37,21 @@ def test_data(self):
...
 
+    def test_create_new_user(self):
+        username = 'test-user'
+        data = {
+            'viewers': [username],
+            'id': None,
+            'name': 'test new catalog',
+            'query': '*',
+        }
+        self.assertEqual(User.objects.filter(username=username).count(), 0)  # pylint: disable=no-member
```
Comment (clintonb) [inline course_discovery/apps/api/tests/test_serializers.py:48]: What's the pylint disable line for?

File: course_discovery/apps/api/serializers.py (hunk 3 above)
Comment (clintonb) [inline course_discovery/apps/api/serializers.py]: A transaction would be preferred. Even if the number is low, I'd prefer that we not have dirty data in the database.

**Comment by clintonb** [comment]
👍  (once `is_valid` concern addressed)

File: course_discovery/apps/api/tests/test_serializers.py (hunk 4 above)
Comment (peter-fogg) [inline course_discovery/apps/api/tests/test_serializers.py:48]: To disable the warning that the user manager has no `filter` member. Pylint doesn't like Django's dynamic property-setting shenanigans.

**Pull request merged**